### ISSUE TITLE
Passing a char** into a record

### ISSUE THREAD

**Issue by rabbiveesh** [Perl5-FFI/FFI-Platypus#182]
In my ongoing quest to wrangle libpostal into FFI, I ran into this wonderful struct:
```C
typedef struct libpostal_normalize_options {
    // List of language codes
    char **languages;  
    size_t num_languages;
    uint16_t address_components;

    // String options
    bool latin_ascii;
    bool transliterate;
    bool strip_accents;
    bool decompose;
    bool lowercase;
    bool trim_string;
    bool drop_parentheticals;
    bool replace_numeric_hyphens;
    bool delete_numeric_hyphens;
    bool split_alpha_from_numeric;
    bool replace_word_hyphens;
    bool delete_word_hyphens;
    bool delete_final_periods;
    bool delete_acronym_periods;
    bool drop_english_possessives;
    bool delete_apostrophes;
    bool expand_numex;
    bool roman_numerals;

} libpostal_normalize_options_t;
```

So the problem is the array of strings at the top there. I tried defining the corresponding record as taking `opaque`, and then planning on casting the values when it's passed to the function that's using them.

BUUUUUT, then I found out that I can't cast a string[] to opaque. So now I'm not sure what to do with this struct.

So in short, how could I go about passing a pointer to an array of strings to a C-land struct?

**Comment by plicease** [comment]
Yeah casting a `string[]` won't work because the array gets free'd before it returns to C space.

You can cast it to a record though:

```perl
my $ffi = FFI::Platypus->new( api => 1 );
my @perl_array = qw( foo bar baz );
my $ptr_size = $ffi->sizeof('opaque');
my $array_size = $ptr_size * @perl_array;
my $c_array = $ffi->cast( 'string[]' => "record($ptr_size)*", \@perl_array);
my $opaque = $ffi->cast( "record($ptr_size)*" => 'opaque', $c_array);
```

It's important that both `@perl_array` and `$c_array` remain in scope and unmodified while `$opaque` is in use.

This should probably be abstracted into a class to avoid insanity.

Also we should probably support pointers to arrays of strings (and other types perhaps) as they are fairly common.  The exact way this should be invoked isn't entirely clear atm because `string[3]` for example already has meaning and `string[]` as a member type may be confusing.

**Comment by plicease** [comment]
I wrote a class that abstracts most (but not all) of the complexity for this:

https://metacpan.org/pod/release/PLICEASE/FFI-Platypus-Record-StringArray-0.01/lib/FFI/Platypus/Record/StringArray.pm

You just have to make sure the F::P::R::StringArray instance remains in scope while it is in use by C.
